extensions/blocks/opentable/edit.js

```diff
@@ -0,0 +1,293 @@
+
+	if ( ! isEqual( validatedAttributes, attributes ) ) {
+		setAttributes( validatedAttributes );
+	}
+
+	const { rid, style, iframe, domain, lang, newtab } = attributes;
+	const [ notice, setNotice ] = useState();
+
+	const setErrorNotice = () =>
+		setNotice(
+			<>
+				<strong>{ __( 'We ran into an issue', 'jetpack' ) }</strong>
+				<br />
+				{ __( 'Please ensure this embed matches the one from your OpenTable account', 'jetpack' ) }
+			</>
+		);
+
+	const parseEmbedCode = embedCode => {
+		if ( ! embedCode ) {
+			setErrorNotice();
+			return;
+		}
+
+		const scriptTagAttributes = embedCode.match( /< *script[^>]*src *= *["']?([^"']*)/i );
+		if ( ! scriptTagAttributes || ! scriptTagAttributes[ 1 ] ) {
+			setErrorNotice();
+			return;
+		}
+
+		let src = '';
+		if ( scriptTagAttributes[ 1 ].indexOf( 'http' ) === 0 ) {
+			src = new URL( scriptTagAttributes[ 1 ] );
+		} else {
+			src = new URL( 'http:' + scriptTagAttributes[ 1 ] );
+		}
+
+		if ( ! src.search ) {
+			setErrorNotice();
+			return;
+		}
+
+		const searchParams = new URLSearchParams( src.search );
+		let styleSetting = searchParams.get( 'theme' );
+		if ( searchParams.get( 'type' ) === 'button' ) {
+			styleSetting = searchParams.get( 'type' );
+		}
+
+		const newAttributes = {
+			rid: searchParams.getAll( 'rid' ),
+			iframe: Boolean( searchParams.get( 'iframe' ) ),
+			domain: searchParams.get( 'domain' ),
+			lang: searchParams.get( 'lang' ),
+			newtab: Boolean( searchParams.get( 'newtab' ) ),
+			style: styleSetting,
+		};
+
+		const validatedNewAttributes = getValidatedAttributes( defaultAttributes, newAttributes );
+		setAttributes( validatedNewAttributes );
+	};
+
+	const styleOptions = getStyleOptions( rid );
+	const styleValues = getStyleValues( rid );
+
+	const updateStyle = newStyle => {
+		setAttributes( { style: newStyle } );
+	};
+
+	const getTypeAndTheme = fromStyle =>
+		rid.length > 1
+			? [ 'multi', 'button' !== fromStyle ? fromStyle : 'standard' ]
+			: [
+					'button' === fromStyle ? 'button' : 'standard',
+					'button' === fromStyle ? 'standard' : fromStyle,
+			  ];
+
+	const blockPreview = styleOveride => {
+		const [ type, theme ] = getTypeAndTheme( styleOveride ? styleOveride : style );
+		return (
+			<>
+				<div className={ `${ className }-overlay` }></div>
+				<iframe
+					title={ `Open Table Preview ${ clientId }` }
+					src={ `https://www.opentable.com/widget/reservation/canvas?rid=${ join(
+						rid,
+						'%2C'
+					) }&type=${ type }&theme=${ theme }&overlay=false&domain=${ domain }&lang=${
+						lang && languageValues.includes( lang ) ? lang : 'en-US'
+					}&newtab=${ newtab }&disablega=true` }
+				/>
+			</>
+		);
+	};
+
+	const blockControls = (
+		<BlockControls>
+			{ ! isEmpty( rid ) && (
+				<Toolbar
+					isCollapsed={ true }
+					icon="edit"
+					label={ __( 'Type', 'jetpck' ) }
+					controls={ styleOptions.map( styleOption => ( {
```

Comment:
This feels a bit out of place to me.
As far as I know, no other block with style options let the user change it from the toolbar.
I'm not adverse to adding more functionality to the toolbar instead of the sidebar, but in this particular case, the pencil icon is (or maybe _was_? 🤔) used to toggle between placeholder and preview (see for example the Image block, with the relatively new swap/edit icon, which was a pencil once).

I guess we could do a few things here:
- Change the `edit` icon into something more style-related ([`admin-appearance`](https://developer.wordpress.org/resource/dashicons/#admin-appearance)?).
- Add a ["swap"](https://github.com/WordPress/gutenberg/blob/2c1e3be7d1cec76176db6cccee1c6add93cc3e4d/packages/block-library/src/image/icon.js#L8) button that reverts the block back to the placeholder state with the empty restaurant picker field.
- Remove the Embed Code section from the sidebar (its purpose would be resolved by the "swap" button).

Reply:
I think this is how real style options work - they appear in both places - try it with an image block...

Reply:
@scruffian I don't see the Image styles in the toolbar, but only in the sidebar... 😮

Reply:
Its hard to spot:
<img width="562" alt="Screenshot 2020-01-06 at 17 36 58" src="https://user-images.githubusercontent.com/275961/71836120-3aaa4680-30ab-11ea-989b-29f06efffcb3.png">

Reply:
TIL!
I had no idea that Core was overloading the transform button with block styles too!

This said, it's still slightly different than the OpenTable block.
I think that UI comes for free when you use the "vanilla block styles approach" (I've never coded them, so I'm totally unfamiliar with how they work), which looks slightly different than what we have done here.
I'll see if there's a way to simplify and make it more consistent with Core Guten.

Reply:
Yeah, I share your concern. Happy to be lead by @lessbloat...

Reply:
Yeah, a couple of things:

- I think we're trying to move away from the situation where clicking the pencil takes you back to the blank slate state.
- Typically with style variants, we do show them both in the right sidebar, and in the transform dropdown, but these are not true style variants. We'd need to submit a patch in core to make these show up in the transform menu. I'm not against doing that, it will just mean that this will take longer to ship. If we decide to go that route, I'd prefer to do this in three phases: 1) Ship what we've got, 2) Make the patch in core, 3) Circle back and update this block.
- The pencil icon is a bit of a hack, but I like having it there. Otherwise on smaller resolutions, it's not clear that you can change the style of this embed.

**Our options**

A) Keep this as is. Maintain both the style options on the right and the style change option via the pencil icon dropdown in the toolbar.

B) Work through the 3 phases that I outlined above to make it possible to show these style variations in the transform menu. I'm okay proceeding with this option. We have no guarantee that it will be approved in core though.

C) Drop the pencil icon dropdown. Just show the style change in the right sidebar. My worry with this option is discoverability at lower screen resolutions.

D) Remove the pencil icon dropdown, then convert the current right sidebar style selector to a simple dropdown. My worry with this option is discoverability. I think many people will just gloss over the dropdown and not know that it's there.